cpp/arcticdb/storage/rocksdb/rocksdb_storage.cpp

```diff
@@ -0,0 +1,239 @@
+/* Copyright 2023 Man Group Operations Limited
+ *
+ * Use of this software is governed by the Business Source License 1.1 included in the file licenses/BSL.txt.
+ *
+ * As of the Change Date specified in that file, in accordance with the Business Source License, use of this software will be governed by the Apache License, version 2.0.
+ */
+
+#include <arcticdb/storage/rocksdb/rocksdb_storage.hpp>
+
+#include <filesystem>
+
+#include <arcticdb/util/preconditions.hpp>
+#include <arcticdb/entity/performance_tracing.hpp>
+#include <arcticdb/storage/storage_options.hpp>
+#include <arcticdb/storage/storage_utils.hpp>
+#include <arcticdb/entity/serialized_key.hpp>
+#include <arcticdb/codec/segment.hpp>
+
+#include <rocksdb/db.h>
+#include <rocksdb/options.h>
+#include <rocksdb/utilities/options_util.h>
+#include <rocksdb/slice.h>
+
+namespace arcticdb::storage::rocksdb {
+
+namespace fs = std::filesystem;
+namespace fg = folly::gen;
+
+RocksDBStorage::RocksDBStorage(const LibraryPath &library_path, OpenMode mode, const Config& conf) :
+    Storage(library_path, mode) {
+
+    fs::path root_path = conf.path().c_str();
+    auto lib_path_str = library_path.to_delim_path(fs::path::preferred_separator);
+
+    auto lib_dir = root_path / lib_path_str;
+    auto db_name = lib_dir.generic_string();
+
+    std::set<std::string> key_names;
+    arcticdb::entity::foreach_key_type([&](KeyType&& key_type) {
+        std::string handle_name = fmt::format("{}", key_type);
+        key_names.insert(handle_name);
+    });
+
+    std::vector<::rocksdb::ColumnFamilyDescriptor> column_families;
+    ::rocksdb::DBOptions db_options;
+
+    ::rocksdb::ConfigOptions cfg_opts; // not used
+    auto s = ::rocksdb::LoadLatestOptions(cfg_opts, db_name, &db_options, &column_families);
+    if (s.ok()) {
+        std::set<std::string> existing_key_names{};
+        for (const auto& desc : column_families) {
+            if (desc.name != ::rocksdb::kDefaultColumnFamilyName) {
+                existing_key_names.insert(desc.name);
+            }
+        }
+        util::check(existing_key_names == key_names, "Existing database has incorrect key columns.");
+    } else if (s.IsNotFound()) {
+        util::check_arg(mode > OpenMode::READ, "Missing dir {} for lib={}. mode={}",
+                            db_name, lib_path_str, mode);
+        // Default column family required, error if not provided.
+        column_families.emplace_back(::rocksdb::kDefaultColumnFamilyName, ::rocksdb::ColumnFamilyOptions());
+        for (const auto& key_name: key_names) {
+            util::check(key_name != ::rocksdb::kDefaultColumnFamilyName,
```

Comment:
I've put the [`::rocksdb::kDefaultColumnFamilyName`](https://github.com/facebook/rocksdb/blob/main/db/db_impl/db_impl.cc#L113) (which is `"default"`) in the error message now. This isn't really a user-facing error though, it will only happen at run-time if we add a new key type with the name `"default"`, even then we use the key type short names which are single letter, e.g. `"T"` etc., so there really shouldn't ever be a clash.